File: recursion/isPalindrome.js
/*
  - use pointers to check first and last
    - pointers are passed in, otherwise default to 0 and str.length - 1

  - if str.length === 0 or 1, return true
  - check if first === last
    - if it does, then recurse
    - o/w return false

  Time: O(N)
  Space: O(N)
*/
function isPalindrome(str, first=0, last=str.length - 1) {
  if (first >= last) return true;
  if (str[first] === str[last]) return isPalindrome(str, first + 1, last - 1);

  return false;
}

console.log(isPalindrome('abccba') === true);
console.log(isPalindrome('a') === true);
console.log(isPalindrome('abc') === false);
console.log(isPalindrome('abccb') === false);
console.log(isPalindrome('   ') === true);

File: recursion/is_palindrome.rb
def is_palindrome?(str, first=0, last=str.length - 1)
  return true if first >= last
  return is_palindrome??(str, first + 1, last - 1) if str[first] == str[last]

  false
end

p is_palindrome?('abccba') === true
p is_palindrome?('a') === true
p is_palindrome?('abc') === false
p is_palindrome?('abccb') === false
p is_palindrome?('   ') === true
